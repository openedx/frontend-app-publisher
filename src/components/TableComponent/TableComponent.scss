/* stylelint-disable */
thead,
thead tr,
tbody {
  display: block;
}

tbody {
  height: auto;
  overflow-y: auto;
  overflow-x: hidden;

  &::-webkit-scrollbar {
    -webkit-appearance: none;

    &:vertical {
      width: 11px;
    }
    &:horizontal {
      height: 11px;
    }
  }

  &::-webkit-scrollbar-thumb {
    border-radius: 8px;
    border: 2px solid #ffffff;
    background-color: rgba(0, 0, 0, .5);
  }

  &::-webkit-scrollbar-track {
    background-color: #fff;
    border-radius: 8px;
  }
}

td,
th {
  max-width: 25vw;
  width: 25vw;
}

/* Manually override data table max width for td to accomodate full width size*/ 
td.pgn__data-table-cell-wrap {
    max-width: 25vw;
    word-break: break-all;
}
