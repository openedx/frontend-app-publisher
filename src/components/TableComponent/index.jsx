import React from 'react';
import PropTypes from 'prop-types';
import qs from 'query-string';
import { Pagination, Table } from '@edx/paragon';
import 'font-awesome/css/font-awesome.css';

import './TableComponent.scss';

import LoadingSpinner from '../LoadingSpinner';
import StatusAlert from '../StatusAlert';
import { getErrorMessages, getPageOptionsFromUrl, updateUrl } from '../../utils';

class TableComponent extends React.Component {
  componentDidMount() {
    // Get initial data
    this.props.paginateTable();
    this.props.fetchEditorFilterOptions();
  }

  componentDidUpdate(prevProps) {
    const { location } = this.props;

    // Handle the case where the query params have changed. This is used when sorting & paging, but
    // also when the back button is used. We need to determine if this is a pagination or sorting
    // request as we handle these as slightly different actions in the action handlers.
    if (location.search !== prevProps.location.search) {
      const {
        page: prevPage,
        ordering: prevOrdering,
        filter: prevFilter,
        editors: prevEditors,
        course_run_statuses: prevCourseRunStatuses,
      } = qs.parse(prevProps.location.search);
      const {
        page,
        ordering,
        filter,
        editors,
        course_run_statuses: courseRunStatuses,
      } = qs.parse(location.search);
      if (ordering !== prevOrdering) {
        this.props.sortTable(ordering);
      } else if (page !== prevPage) {
        this.props.paginateTable(parseInt(page, 10));
      } else if (filter !== prevFilter) {
        this.props.filterTable({ pubq: filter });
      } else if (editors !== prevEditors) {
        this.props.filterTable(editors);
      } else if (courseRunStatuses !== prevCourseRunStatuses) {
        this.props.filterTable(courseRunStatuses);
      }
    }
  }

  componentWillUnmount() {
    this.props.clearTable();
  }

  renderTableContent() {
    const {
      className,
      pageCount,
      tableSortable,
      data,
      formatData,
      loading,
      location,
    } = this.props;

    const columnConfig = this.props.columns.map(column => ({
      ...column,
      onSort: column.columnSortable ? direction => updateUrl({
        page: 1,
        ordering: direction === 'desc' ? `-${column.key}` : column.key,
      }) : null,
    }));

    let sortDirection;
    let sortColumn;

    if (tableSortable) {
      const { ordering } = qs.parse(location.search);
      sortDirection = ordering && ordering.indexOf('-') !== -1 ? 'desc' : 'asc';
      sortColumn = (ordering && ordering.replace('-', '')) || columnConfig[0].key;
    }

    const paginationOptions = getPageOptionsFromUrl();

    return (
      <div className={className}>
        {loading && <LoadingSpinner />}
        <div className="table-responsive">
          <Table
            className="table-sm table-striped"
            columns={columnConfig}
            data={formatData(data)}
            tableSortable={tableSortable}
            defaultSortedColumn={sortColumn}
            defaultSortDirection={sortDirection}
          />
        </div>
        { pageCount > 1
          && (
          <div className="mt-2 d-flex justify-content-center">
            <Pagination
              pageCount={pageCount}
              currentPage={paginationOptions.page}
              onPageSelect={page => updateUrl({ page })}
            />
          </div>
          )}
      </div>
    );
  }

  renderLoadingMessage() {
    return <LoadingSpinner />;
  }

  renderErrorMessage() {
    const { error } = this.props;
    const isForbidden = error.response && error.response.status === 403;
    const message = isForbidden
      ? ['You do not yet have access to Publisher. Please contact your project coordinator to get access.']
      : ['Unable to load data: '].concat(getErrorMessages(this.props.error));
    return (
      <StatusAlert
        alertType="danger"
        iconClassNames={['fa', 'fa-times-circle']}
        message={message}
      />
    );
  }

  renderEmptyDataMessage() {
    return (
      <StatusAlert
        alertType="warning"
        iconClassNames={['fa', 'fa-exclamation-circle']}
        message="There are no results."
      />
    );
  }

  render() {
    const {
      data,
      loading,
      error,
    } = this.props;

    return (
      <>
        {error && this.renderErrorMessage()}
        {loading && !data && this.renderLoadingMessage()}
        {!loading && !error && data && data.length === 0
          && this.renderEmptyDataMessage()}
        {data && data.length > 0 && this.renderTableContent()}
      </>
    );
  }
}

TableComponent.propTypes = {
  // Props expected from consumer
  className: PropTypes.string,
  columns: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  formatData: PropTypes.func.isRequired,
  tableSortable: PropTypes.bool,

  // Props expected from TableContainer / redux store
  data: PropTypes.arrayOf(PropTypes.shape({})),
  pageCount: PropTypes.number.isRequired,
  loading: PropTypes.bool,
  error: PropTypes.instanceOf(Error),
  paginateTable: PropTypes.func.isRequired,
  sortTable: PropTypes.func.isRequired,
  filterTable: PropTypes.func.isRequired,
  clearTable: PropTypes.func.isRequired,
  location: PropTypes.shape({
    search: PropTypes.string,
  }).isRequired,
  fetchEditorFilterOptions: PropTypes.func.isRequired,
};

TableComponent.defaultProps = {
  className: '',
  tableSortable: false,
  data: undefined,
  error: null,
  loading: false,
};

export default TableComponent;
