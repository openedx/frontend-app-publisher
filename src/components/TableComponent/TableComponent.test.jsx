import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import TableComponent from './index';

describe('CourseTable', () => {
  const mockFormat = data => data.map(datum => ({
    ...datum,
  }));
  const mockColumns = [
    {
      label: 'Col 1',
      key: 'col1',
      columnSortable: true,
    },
    {
      label: 'Col 2',
      key: 'col2',
      columnSortable: false,
    },
  ];
  const mockFetch = () => [
    {
      col1: 'Test11',
      col2: 'Test12',
    },
    {
      col1: 'Test21',
      col2: 'Test22',
    },
  ];

  it('shows a table', () => {
    const component = shallow(<TableComponent
      id="test"
      className="test"
      fetchMethod={mockFetch}
      columns={mockColumns}
      formatData={mockFormat}
      tableSortable
      pageCount={1}
      itemCount={2}
      paginateTable={() => true}
      sortTable={() => true}
      clearTable={() => true}
      location={{ search: '?page=1&ordering=key' }}
      fetchEditorFilterOptions={() => true}
    />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('shows an error', () => {
    const component = shallow(<TableComponent
      id="test"
      className="test"
      fetchMethod={mockFetch}
      columns={mockColumns}
      formatData={mockFormat}
      error={new Error('Test error')}
      tableSortable
      pageCount={1}
      paginateTable={() => true}
      sortTable={() => true}
      clearTable={() => true}
      location={{ search: '?page=1&ordering=key' }}
      fetchEditorFilterOptions={() => true}
    />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('shows loading', () => {
    const component = shallow(<TableComponent
      id="test"
      className="test"
      fetchMethod={mockFetch}
      columns={mockColumns}
      formatData={mockFormat}
      tableSortable
      pageCount={1}
      paginateTable={() => true}
      sortTable={() => true}
      clearTable={() => true}
      location={{ search: '?page=1&ordering=key' }}
      loading
      fetchEditorFilterOptions={() => true}
    />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('shows an empty table', () => {
    const component = shallow(<TableComponent
      id="test"
      className="test"
      fetchMethod={mockFetch}
      columns={mockColumns}
      formatData={mockFormat}
      tableSortable
      pageCount={1}
      paginateTable={() => true}
      sortTable={() => true}
      clearTable={() => true}
      location={{ search: '?page=1&ordering=key' }}
      data={[]}
      fetchEditorFilterOptions={() => true}
    />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('shows a populated table', () => {
    const component = shallow(<TableComponent
      id="test"
      className="test"
      fetchMethod={mockFetch}
      columns={mockColumns}
      formatData={mockFormat}
      tableSortable
      pageCount={1}
      itemCount={2}
      paginateTable={() => true}
      sortTable={() => true}
      clearTable={() => true}
      location={{ search: '?page=1&ordering=key' }}
      data={mockFetch()}
      fetchEditorFilterOptions={() => true}
    />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('shows a populated table after a component update for page', () => {
    const component = shallow(<TableComponent
      id="test"
      className="test"
      fetchMethod={mockFetch}
      columns={mockColumns}
      formatData={mockFormat}
      tableSortable
      pageCount={1}
      itemCount={2}
      paginateTable={() => true}
      sortTable={() => true}
      clearTable={() => true}
      location={{ search: '?page=1&ordering=key' }}
      data={mockFetch()}
      fetchEditorFilterOptions={() => true}
    />);
    component.setProps({
      location: {
        search: '?page=2&ordering=key',
      },
    });
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('shows a populated table after a component update for ordering', () => {
    const component = shallow(<TableComponent
      id="test"
      className="test"
      fetchMethod={mockFetch}
      columns={mockColumns}
      formatData={mockFormat}
      tableSortable
      pageCount={1}
      itemCount={2}
      paginateTable={() => true}
      sortTable={() => true}
      clearTable={() => true}
      location={{ search: '?page=1&ordering=key' }}
      data={mockFetch()}
      fetchEditorFilterOptions={() => true}
    />);
    component.setProps({
      location: {
        search: '?page=1&ordering=-key',
      },
    });
    expect(shallowToJson(component)).toMatchSnapshot();
  });
});
