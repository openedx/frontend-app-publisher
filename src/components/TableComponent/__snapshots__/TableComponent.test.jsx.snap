// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseTable shows a populated table 1`] = `
<Fragment>
  <div
    className="test"
  >
    <div
      className="table-responsive"
    >
      <withDeprecatedProps(Table)
        className="table-sm table-striped"
        columns={
          Array [
            Object {
              "columnSortable": true,
              "key": "col1",
              "label": "Col 1",
              "onSort": [Function],
            },
            Object {
              "columnSortable": false,
              "key": "col2",
              "label": "Col 2",
              "onSort": null,
            },
          ]
        }
        data={
          Array [
            Object {
              "col1": "Test11",
              "col2": "Test12",
            },
            Object {
              "col1": "Test21",
              "col2": "Test22",
            },
          ]
        }
        defaultSortDirection="asc"
        defaultSortedColumn="key"
        tableSortable={true}
      />
    </div>
  </div>
</Fragment>
`;

exports[`CourseTable shows a populated table after a component update for ordering 1`] = `
<Fragment>
  <div
    className="test"
  >
    <div
      className="table-responsive"
    >
      <withDeprecatedProps(Table)
        className="table-sm table-striped"
        columns={
          Array [
            Object {
              "columnSortable": true,
              "key": "col1",
              "label": "Col 1",
              "onSort": [Function],
            },
            Object {
              "columnSortable": false,
              "key": "col2",
              "label": "Col 2",
              "onSort": null,
            },
          ]
        }
        data={
          Array [
            Object {
              "col1": "Test11",
              "col2": "Test12",
            },
            Object {
              "col1": "Test21",
              "col2": "Test22",
            },
          ]
        }
        defaultSortDirection="desc"
        defaultSortedColumn="key"
        tableSortable={true}
      />
    </div>
  </div>
</Fragment>
`;

exports[`CourseTable shows a populated table after a component update for page 1`] = `
<Fragment>
  <div
    className="test"
  >
    <div
      className="table-responsive"
    >
      <withDeprecatedProps(Table)
        className="table-sm table-striped"
        columns={
          Array [
            Object {
              "columnSortable": true,
              "key": "col1",
              "label": "Col 1",
              "onSort": [Function],
            },
            Object {
              "columnSortable": false,
              "key": "col2",
              "label": "Col 2",
              "onSort": null,
            },
          ]
        }
        data={
          Array [
            Object {
              "col1": "Test11",
              "col2": "Test12",
            },
            Object {
              "col1": "Test21",
              "col2": "Test22",
            },
          ]
        }
        defaultSortDirection="asc"
        defaultSortedColumn="key"
        tableSortable={true}
      />
    </div>
  </div>
</Fragment>
`;

exports[`CourseTable shows a table 1`] = `<Fragment />`;

exports[`CourseTable shows an empty table 1`] = `
<Fragment>
  <StatusAlert
    alertType="warning"
    className=""
    dismissible={false}
    iconClassNames={
      Array [
        "fa",
        "fa-exclamation-circle",
      ]
    }
    message="There are no results."
    onClose={[Function]}
    title={null}
  />
</Fragment>
`;

exports[`CourseTable shows an error 1`] = `
<Fragment>
  <StatusAlert
    alertType="danger"
    className=""
    dismissible={false}
    iconClassNames={
      Array [
        "fa",
        "fa-times-circle",
      ]
    }
    message={
      Array [
        "Unable to load data: ",
        "Test error.",
      ]
    }
    onClose={[Function]}
    title={null}
  />
</Fragment>
`;

exports[`CourseTable shows loading 1`] = `
<Fragment>
  <LoadingSpinner
    message="Loadingâ€¦"
  />
</Fragment>
`;
